### PR TITLE
Embed: design doc for new embed API

### PR DESCRIPTION
**Pull request by stsewd**
Mostly based on my notes from https://github.com/readthedocs/readthedocs.org/issues/7963.

### CODE REVIEW

File: docs/development/design/embed-api.rst [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+Current implementation
+----------------------
+
+The current implementation of the API is partially documented in :doc:`/guides/embedding-content`.
+Some characteristics/problems are:
+
+- There are three ways of querying the API, and some rely on Sphinx's concepts like ``doc``.
+- Doesn't cache responses or doesn't purge the cache on build.
+- Doesn't support MkDocs.
+- It returns all sections from the current page.
```
Comment (humitos) [inline docs/development/design/embed-api.rst:24]: You can send `section=` to get only the section you want.

File: docs/development/design/embed-api.rst [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+- It doesn't return valid HTML for definition lists (``dd`` tags without a ``dt`` tag).
+- The client doesn't know if the page requires extra JS or CSS in order to make it work or look nice.
+
+Improvements
+------------
+
+These improvements aren't breaking changes, so we can implement them in the old and new API.
+
+- Support for MkDocs.
+- Always return a valid/well formed HTML block.
```
Comment (humitos) [inline docs/development/design/embed-api.rst:38]: If this refers to the `dd`/`dt` issue, I'm not sure if this will be easily doable for all the cases. We may end up re-parsing the `dd` to convert that HTML tag to another one, but that will break the style based on `dd` tags. So, this will be tricky to do it right and make it work for all the themes --which is also related with the new `extras` field.

File: docs/development/design/embed-api.rst [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+-------
+
+The API would be split into two endpoints, and only have one way of querying the API.
+
+Get page
+--------
+
+Allow us to query information about a page, like its list of sections.
+
+.. http:get:: /_/api/v3/embed/pages?project=docs&version=latest&path=install.html
```
Comment (humitos) [inline docs/development/design/embed-api.rst:50]: If we will have only one way of querying the API, we should probably use `?url=` instead of sending multiple attributes. The implementation of the extensions is easier (and there are things not possible to implement without it --e.g. intersphinx) and the usage of the endpoint will be simplified too.

File: docs/development/design/embed-api.rst [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+         "sections": [
+            {
+               "title": "Installation",
+               "id": "installation"
+            },
+            {
+               "title": "Examples",
+               "id": "examples"
+            }
+         ],
```
Comment (humitos) [inline docs/development/design/embed-api.rst:73]: I like the idea of listing all the sections. I'd also like to see a way to explore/surf the API just by clicking these links, similar as we do with the `_links` field in other endpoints. This way, you can discover the API by clicking on those, but also expose these links to developers.

File: docs/development/design/embed-api.rst [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+We should have a section in our docs instead of guide where the embed API is documented.
+There we can list v2 as deprecated.
+We would need to migrate our extension as well.
+Most of the parsing code could be shared between the two APIs, so it shouldn't be a burden to maintain.
+
+API Client
+----------
+
+Do we really need a JS client?
```
Comment (humitos) [inline docs/development/design/embed-api.rst:148]: I think so. IIRC @agjohnson idea here was to include the popup/modal form of this into the JS client. Besides, we could eventually be able to use this JS client from our extension itself, instead of maintaining the front-end in our own extension (or adding it as another popup extension together with the ones we currently support)

**Review by humitos — COMMENTED** [review]
I think there are good ideas here.

I don't think we should have 2 separate endpoints for this and we should only rely on `?url=` attribute unless there is a good reason to not to.

Also, we should support `nparagraph=` or `nwords=` or both to be able to chunk the content of the section returned (see https://github.com/readthedocs/readthedocs-ext/pull/304)

File: docs/development/design/embed-api.rst (hunk 1 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:24]: What I mean here is that it always returns all the sections (title and id), not the section content.

File: docs/development/design/embed-api.rst (hunk 2 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:38]: We just need to extract only that element. For example

```html
<dl class="foo">
 .. more..
 <dt>foo</dt> <dd>bar</dd>
.. more...
</dl>
```

becomes

```html
<dl class="foo">
 <dt>foo</dt> <dd>bar</dd>
</dl>
```

Extracting content isn't hard to do here.

File: docs/development/design/embed-api.rst (hunk 3 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:50]: Using a URL makes it depend on the domain, if the domain changes everything is broken. I'm fine supporting both to support intersphinx.

**Comment by stsewd** [comment]
> Also, we should support nparagraph= or nwords= or both to be able to chunk the content of the section returned 

Feels like that's a client thing, to be able to cut the content exactly where they want and show a "See more" link.

File: docs/development/design/embed-api.rst (hunk 2 above)
Comment (humitos) [reply docs/development/design/embed-api.rst:38]: Hrm... I think that has the same problem.

When parsing `citation`/`glossary` we don't need to return the title/definition (`dt`), we only want the description (`dd`). Example in code https://github.com/readthedocs/readthedocs.org/blob/eb682d8b8f480e96e7478d968aa12e52f4118063/readthedocs/embed/views.py#L329-L335

Following, your example, to make it valid it should be:

```html
<dl class="foo">
 <dt></dt>
 <dd>bar</dd>
</dl>
```

However, this won't probably render correctly inside the tooltip, since in that context is not a description list --it's not a list at all-- but just the definition of one of the term from that list.

In those cases, `citation`/`glossary` will be rendered in the same than any other definition list and I'm not sure we want that. We need to think how we will support this. Example:

### Glossary (only includes the definition)

Screenshot_2021-03-30_11-26-04


### Config value (includes the definition list completely repeating the term hovered)

Screenshot_2021-03-30_11-23-06

Comment (stsewd) [reply docs/development/design/embed-api.rst:38]: I think we should repeat the title, it doesn't necessary match the text where the tooltip is present. Removing the title if it's the same as the text can be done in the client side.

File: docs/development/design/embed-api.rst [hunk 6]
```diff
@@ -0,0 +1,154 @@
+Embed API
+=========
+
+The embedded API allows to embed content from docs pages in other sites.
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:4]:
```suggestion
The Embed API allows to embed content from docs pages in other sites.
```

File: docs/development/design/embed-api.rst (hunk 3 above)
Comment (ericholscher) [reply docs/development/design/embed-api.rst:50]:
```suggestion
.. http:get:: /_/api/v3/embed/page/?project=docs&version=latest&path=install.html
```

File: docs/development/design/embed-api.rst [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            "css": ["https://docs.readthedocs.io/en/latest/index.css"],
+         }
+      }
+
+Get section
+-----------
+
+Allow us to query the content of the section, with all links re-written as absolute.
+
+.. http:get:: /_/api/v3/embed/sections?project=docs&version=latest&path=install.html#examples
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:85]:
```suggestion
.. http:get:: /_/api/v3/embed/section/?project=docs&version=latest&path=install.html#examples
```

File: docs/development/design/embed-api.rst [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            },
+            {
+               "title": "Examples",
+               "id": "examples"
+            }
+         ],
+         "extras": {
+            "js": ["https://docs.readthedocs.io/en/latest/index.js"],
+            "css": ["https://docs.readthedocs.io/en/latest/index.css"],
+         }
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:77]: Where does this data come from? Seems hard to parse it properly, and is providing a promise we can't meaningfully keep if we're expecting the CSS to work in a third-party page.

File: docs/development/design/embed-api.rst [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+Notes
+-----
+
+- If a section or page doesn't exist, we return 404.
+- All links are re-written to be absolute (this is already done).
+- All sections listed are from html tags that are linkeable, this is, they have an ``id``
+  (we don't rely on the toctree from the fjson file anymore).
+- The IDs returned don't contain the redundant ``#`` symbol.
+- The content is an string with a well formed HTML block.
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:115]: How are we guaranteeing this?

File: docs/development/design/embed-api.rst [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+- We could also support only ``url`` as argument for ``/sections`` and ``/pages``,
+  but this introduces another way of querying the API.
+  Having two ways of querying the API makes it *less-cacheable*.
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:118]: I'm pretty sure we only want to support URL. One of the goals of the grant work is to support externally hosted Sphinx docs with the embed API, so that would require supporting URL's.

File: docs/development/design/embed-api.rst [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+- Returning the extra js and css requires parsing the HTML page itself,
+  rather than only the content extracted from the fjson files (this is for sphinx).
+  We can use both, the html file and the json file, but we could also just start parsing the full html page
+  (we can re-use code from the search parsing to detect the main content).
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:122]: Yea, I think parsing the content in production is probably correct. The json approach doesn't scale across doc tools or externally hosted stuff, so I think using the canonical document the user is referencing will reduce a lot of complexity (while adding some around parsing, but we already have to do this as you noted in search)

File: docs/development/design/embed-api.rst [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+- We could not return the ``title`` of the page/section as it would require more parsing to do
+  (but we can re-use the code from search).
+  Titles can be useful to build an UI like https://readthedocs.org/projects/docs/tools/embed/.
+- MkDocs support can be added easily as we make our parsing code more general.
+
+.. note::
+
+   We should probably make a distinction between our general API that handles Read the Docs resources,
+   vs our APIs that expose features (like server side search, footer, and embed, all of them proxied).
+   This way we can version each endpoint separately.
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:135]: I don't quite understand this note. You mean API's that are in doc pages vs. our normal REST API?

File: docs/development/design/embed-api.rst (hunk 5 above)
Comment (ericholscher) [reply docs/development/design/embed-api.rst:148]: Yea, the JS client is going to be doing a lot of work in addition to just doing the GET. We definitely want nice default widgets, along with helper methods that make this a much nicer UX for users.

File: docs/development/design/embed-api.rst [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+- Returning the extra js and css requires parsing the HTML page itself,
+  rather than only the content extracted from the fjson files (this is for sphinx).
+  We can use both, the html file and the json file, but we could also just start parsing the full html page
+  (we can re-use code from the search parsing to detect the main content).
+- ``extras`` could be returned only on ``/pages``, or only on ``/sections``.
+  It makes more sense to be only on ``/pages``,
+  but them querying a section would require to query a page to get the extra js/css files.
+- We could not return the ``title`` of the page/section as it would require more parsing to do
+  (but we can re-use the code from search).
+  Titles can be useful to build an UI like https://readthedocs.org/projects/docs/tools/embed/.
```
Comment (ericholscher) [inline docs/development/design/embed-api.rst:128]: How would a user build something like this without the ability to also get _all_ the pages in a project? We aren't exposing an API for that in this design doc. Should we?

**Review by ericholscher — COMMENTED** [review]
I think this is a good start, but I think we need a lot more thought and explanation behind some of the design decisions. I'd love to see more structure for a design doc here, explicitly stating the goals in a more detailed way. There is a _ton_ of implied context in each "list of 10 items" that we should be explaining more fully. 

Most importantly, we should have a sections of goals, and why they are valuable. For example, why is "Always return a valid/well formed HTML block." important? How are we going to ensure that at the code level?

File: docs/development/design/embed-api.rst (hunk 3 above)
Comment (ericholscher) [reply docs/development/design/embed-api.rst:50]: I'm also not sure if we want to make this a v3 API, given that it works totally differently than the others :/

File: docs/development/design/embed-api.rst (hunk 13 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst]: Hmm, good point, we could have an endpoint for that, but now that you mentioned we need to support external sites, I don't think it would be possible. Or it would be rtd only feature.

File: docs/development/design/embed-api.rst (hunk 12 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:181]: Yeah, basically the APIs from readthedocs.org/api/ vs docs.readthedocs.io/_/api

File: docs/development/design/embed-api.rst (hunk 10 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst]: Didn't know we were required to make it work with external sites, I updated the doc with some thoughts around that.

File: docs/development/design/embed-api.rst (hunk 9 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst]: I just expanded this in the new update

File: docs/development/design/embed-api.rst (hunk 8 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:72]: I just expanded this in the new update

File: docs/development/design/embed-api.rst (hunk 3 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:45]: > I'm also not sure if we want to make this a v3 API, given that it works totally differently than the others :/

That's why my note in https://github.com/readthedocs/readthedocs.org/pull/8039#discussion_r604452865

File: docs/development/design/embed-api.rst (hunk 7 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:80]: I'm not sure about the naming conventions here. We are naming a resource, so it should be plural, but we only have a single resource to get the data from.

File: docs/development/design/embed-api.rst (hunk 8 above)
Comment (humitos) [reply docs/development/design/embed-api.rst:72]: The idea is good to me and could be useful in different cases. However, it's hard to make it work in _all of them_. For example, if the `.css` has relative URLs on its content it will break completely when loaded under a different path.

File: docs/development/design/embed-api.rst [hunk 14]
```diff
@@ -104,36 +99,112 @@ Allow us to query the content of the section, with all links re-written as absol
...
+Return extra js and css that may be required to render the page correctly.
+  We return a list of js and css files that are included in the page ``style`` and ``script`` tags.
+  The returned js and css files aren't guaranteed to be required in order to render the content,
+  but a decision for the client to make. Of course users can also anticipate the kind of content
+  they want to embed and extract the correct css and js in order to make it work.
+  We won't check for inline scripts.
+
+``extras`` could be returned only on ``/pages``, or only on ``/sections``.
   It makes more sense to be only on ``/pages``,
   but them querying a section would require to query a page to get the extra js/css files.
```
Comment (humitos) [inline docs/development/design/embed-api.rst]: `extras` .css and .js are useful for section as well, for example when the section returned contains mathjax and/or sphinx-tabs.

We could expand this section mentioning these cases and thinking if they will work with the structure proposed here. We already have them working on sphinx-hoverxref and we could check with it if it makes sense.

File: docs/development/design/embed-api.rst [hunk 15]
```diff
@@ -104,36 +99,112 @@ Allow us to query the content of the section, with all links re-written as absol
...
+
+  Instead of
+
+  .. code:: html
+
+     <dd>Some definition</dd>
+
+  Note that we only try to keep the current structure,
+  if the page contains malformed HTML, we don't try to *fix it*.
+  This improvement can be shared with the current API (v2).
```
Comment (humitos) [inline docs/development/design/embed-api.rst:159]: We can't back-port this since it's a breaking change from a user's perspective: the render of the content in the tooltip will be completely different (see https://github.com/readthedocs/readthedocs.org/pull/8039#discussion_r603396695)

File: docs/development/design/embed-api.rst [hunk 16]
```diff
@@ -104,36 +99,112 @@ Allow us to query the content of the section, with all links re-written as absol
...
 
    We should probably make a distinction between our general API that handles Read the Docs resources,
    vs our APIs that expose features (like server side search, footer, and embed, all of them proxied).
    This way we can version each endpoint separately.
 
+Support for external sites
+--------------------------
+
+Currently this document uses ``project``, ``version``, and ``path`` to query the API,
+but since the CZI grant requires this to work with external sites, those arguments can be replaced with ``url``.
```
Comment (humitos) [inline docs/development/design/embed-api.rst:187]: We require this to make intersphinx to work, not just because CZI and/or external sites.

File: docs/development/design/embed-api.rst [hunk 17]
```diff
@@ -104,36 +99,112 @@ Allow us to query the content of the section, with all links re-written as absol
...
+
+Considerations
+``````````````
+
+If a project changes its custom domain, current usage of the API would break.
```
Comment (humitos) [inline docs/development/design/embed-api.rst:192]: I think we could check `Referer`/`Origin` header as a fallback to try to find the project in case they changed the domain (sent via `?url=`).

File: docs/development/design/embed-api.rst [hunk 18]
```diff
@@ -104,36 +99,112 @@ Allow us to query the content of the section, with all links re-written as absol
...
+
+We would need to check if the domain belongs to a project inside RTD and fetch the file from storage,
+and if it's from an external site fetch it from the internet.
+
+The API could be missused.
+This is already true if we don't support external sites,
+since we host arbitrary HTML already.
+But it can be abussed to crawl external sites without the consent.
+We can integrate support for external sites in a later stage,
+or have a list of allowed sites.
```
Comment (humitos) [inline docs/development/design/embed-api.rst:202]: I'm not sure to understand what's the problem or abuse we can face here.

I think we definitely don't have a list of allowed sites, here. We want this to be general/global.

File: docs/development/design/embed-api.rst (hunk 8 above)
Comment (humitos) [reply docs/development/design/embed-api.rst:72]: Note that Sphinx 3.5 includes a feature to add JS and CSS files _only to the pages where "they are used"_ and this potentially kills our extension to render dynamic content. See https://github.com/sphinx-doc/sphinx/pull/8631#issuecomment-821155458

So, having a way to know what are the JS _and_ CSS files required in the target page is good.

Comment (stsewd) [reply docs/development/design/embed-api.rst:72]: I was actually thinking to remove this, I put some of my thoughts in the document, basically most users want to inject content they know about or trust, so they know the required CSS/JS to make it work. I think you could refactor the extension to have a list of js/css files to include in all pages and auto-populate that with files from popular extensions based on some settings.

File: docs/development/design/embed-api.rst (hunk 17 above)
Comment (humitos) [reply docs/development/design/embed-api.rst:193]: We could also save all the historic domains for a project. Instead of delete them, we can mark them as inactive.

Comment (ericholscher) [reply docs/development/design/embed-api.rst:193]: That seems very difficult to scale. What if another project reuses the domain?

File: docs/development/design/embed-api.rst (hunk 8 above)
Comment (ericholscher) [reply docs/development/design/embed-api.rst:72]: I think we came down on "Let's remove it for now -- we can always add it later if we find a good use-case for it, but let's not start with it". Specifically, the _client_ should know enough about the content it's embedding to solve this usecase (eg. `if there is a sphinx-tabs classname, add the sphinx-tabs CSS`)

Comment (davidfischer) [reply docs/development/design/embed-api.rst:72]: > Specifically, the client should know enough about the content it's embedding to solve this usecase

:+1:

Also, dynamically adding CSS/JS is a security concern. I don't think there's any way to do that safely.

File: docs/development/design/embed-api.rst (hunk 18 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst]: We would be exposed to [SSRF](https://en.wikipedia.org/wiki/Server-side_request_forgery), basically an attacker can use our servers to make get requests to other insecure sites or DDOS another site.

File: docs/development/design/embed-api.rst (hunk 2 above)
Comment (humitos) [reply docs/development/design/embed-api.rst]: Glossary and Citation special cases can be managed from the frontend/client since we have all the data we need (note that we return the `.next()` HTML element). We will just need to parse it remove the content we don't want to display.

_However_, for the `dt` case (definition of a config, Python method/function, etc) it won't work because we need the `.parent()` of HTML element picked by ID. Take a look at this example:

```html
<dl class="std confval">
  <dt id="confval-hoverxref_role_types">
    <code class="sig-name descname">hoverxref_role_types</code>
    <a class="headerlink" href="#confval-hoverxref_role_types" title="Permalink to this definition">¶</a>
  </dt>
  <dd>
    <p>Description: Style to use by default when hover each type of reference (role).</p>
    <!-- ... snip ... -->
  </dd>
</dl>
```

(example from https://sphinx-hoverxref.readthedocs.io/en/latest/configuration.html#confval-hoverxref_role_types)

With the id `confval-hoverxref_role_types` we are currently returning the whole `dl` because we access to the _parent_ element. If we return just the HTML tag matched, we will be returning only the `dt` which does not include the description nor any content.

Without handling this special case properly, we will show tooltips like:

Screenshot_2021-05-27_16-16-41



We need to find a general solution for these special cases.

File: docs/development/design/embed-api.rst (hunk 3 above)
Comment (humitos) [reply docs/development/design/embed-api.rst:44]: > Using a URL makes it depend on the domain, if the domain changes everything is broken

Yes and no. Maybe (?), I guess. If they use `https://readthedocs.org/api/v3/embed/?url=` endpoint and it shouldn't break if they _don't remove the Domain for that Project_ from our platform. Even if the DNS for that domain is dead.

File: docs/development/design/embed-api.rst [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+         "sections": [
+            {
+               "title": "Installation",
+               "id": "installation"
+            },
+            {
+               "title": "Examples",
+               "id": "examples"
+            }
+         ]
```
Comment (humitos) [inline docs/development/design/embed-api.rst:67]: What's the use case for this field? Are we returning nested sections here or only the first level?

File: docs/development/design/embed-api.rst (hunk 2 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst]: @humitos not sure to understand the problem there? that's the same example we have in the document. The id is on the dt element, so we fetch the dt and the next dd enclosed in the dl tag.

File: docs/development/design/embed-api.rst (hunk 19 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:67]: All sections, nested as well, so the client knows what sections it can query, this is only from h elements, but I guess it could be expanded for more tags.

Comment (humitos) [reply docs/development/design/embed-api.rst:67]: In that case, why only show `h` tags when the backend can reply to any anchor link? If we want expose _all the possible queries_ we should return all the valid anchors.

In any case, I think we should add a request GET argument to expand this and avoid returning a lot of data when it's not going to be used. Example: `?expand=sections`

File: docs/development/design/embed-api.rst (hunk 2 above)
Comment (humitos) [reply docs/development/design/embed-api.rst]: I may not have expressed myself properly.

This is a _special case_ because we are manipulating the HTML to manage a particular case on Sphinx: **return the parent of a particular ID, instead of itself**. Since we are going to support external sites that will not be generated with Sphinx (*), we can't assume that we always have to do this special HTML manipulation. That's why I'm saying that we need a general solution to differentiate these special cases from regular (non-Sphinx) pages.

I'm thinking that we will need to communicate this to the backend somehow. Maybe a request GET argument: `?doctool=sphinx&version=3.5.1`

(*) or even with a different version of Sphinx that may adds the `id=` in the `dl`.

Comment (stsewd) [reply docs/development/design/embed-api.rst]: that's part of how we expect the html to be, following semantic html and using correct tags, if a site has malformed tags or if the id is in a different element we don't try to fix it, we just return it as is.

File: docs/development/design/embed-api.rst (hunk 19 above)
Comment (stsewd) [reply docs/development/design/embed-api.rst:67]: This is in another endpoint, so I don't think we need that feature here.
